obspy/core/utcdatetime.py

```diff
@@ -1502,6 +1458,53 @@ def _set_precision(self, value=6):
 
     precision = property(_get_precision, _set_precision)
 
+    def replace(self, **kwargs):
+        """
+        Return a new UTCDateTime object with one or more parameters replaced.
+
+        The following parameters are supported: year, month, day, julday,
+        hour, minute second, microsecond.
```

Comment:
I feel like this should also support nanoseconds and the precision parameter.

Reply:
I thought about supporting nanoseconds but it may introduce some confusion into the API for two reasons:

1. The `ns` public attribute means nanoseconds since 1970, but the other parameters in the `replace` method are relative to the next higher unit of time (eg hour means hour in current day, not hours since 1970). So should nanoseconds be relative to microseconds, or seconds? If so, the usage of nanosecond and the `ns` attribute are inconsistent. If it means nanoseconds since 1970 there is no need to use it in replace, just create a new instance.

2. The default precision does not support nanoseconds, and the vast majority of users do not change the precision. What should replace do if the nanosecond parameter is used but the precision is the default value? Raise a warning and do nothing?

As for replace with the precision parameter: I think this is doable but there are two hangups. Under the proposal outlined in #2072 the internal integer representation of time is rounded according to precision upon init. This is necessary to remove the "fuzzy" behaviour and all the problems it brings. Because of this, however, if `replace` is used to go from a higher precision to a lower one it may cause some confusing behaviour. For example:

```python
from obspy import UTCDateTime
utc = UTCDateTime(10.222, precision=0)
utc.timestamp  # -> 10  (10.222 was rounded to 10 based on a precision of 0)
utc2 = utc.replace(precision=3)
utc2.timestamp  # -> 10.000 (but the user might expect 10.222) 
```
I suppose we could issue a blanket warning whenever a precision greater than the original is set? 

Also, `replace` was intended to avoid messy inits that could produce the same behaviour. For example:
```python
from obspy import UTCDateTime
utc = UTCDateTime()
# get same utc with hour 10
# using replace
utc2 = utc.replace(hour=10)
# using init
utc2 = UTCDateTime(year=utc.year, month=utc.month, day=utc.day, hour=10, ...)  # and so on
``` 
With replacing a precision, however, it is not unreasonable to simply init a new object. 
```
from obspy import UTCDateTime
utc = UTCDateTime()
utc2 = UTCDateTime(utc, precision=9)
```

Reply:
You clearly gave this a lot more thought than I did and I agree with your reasoning. Thanks a lot for the very comprehensive explanations. Can you add one or two sentences to the docstring to state this in a condensed form?

Reply:
stating this clearly in one or two sentences is the trick... will do my best